feat: add SchemaService and FillEntityDefaults utility

GGabriele opened the pull request:
Mirroring what's been previously done for `Plugins`,
this adds `GetFullSchema` to `Routes`, `Upstreams`, `Targets` and `Services`.
This also adds utils function to fill core entities with
defaults comings from their schema.

hbagdi commented:
Let's do this a bit differently.
Let's create a SchemaService with a function to pull a schema for any entity. `Get(entityname string)` something like. That is much more concise and meets the goal.

GGabriele commented:
@hbagdi good point.

Since I removed the `GetFullSchema` from each entity service, I also moved all the tests under `kong/utils_test.go`. Let me know what you think.

hbagdi commented on kong/schema_service.go:
```diff
@@ -0,0 +1,29 @@
+type SchemaService service
+
+// Get retrieves the full schema of kong entities.
+func (s *SchemaService) Get(ctx context.Context, entity string) (map[string]interface{}, error) {
```
make a type 
```
type map[string]interface{} Schema
```
return that here and then take this as input to the util package. Might result in import cycles.

hbagdi commented on kong/schema_service_test.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGet(T *testing.T) {
```
this should follow the naming convention found in other service tests.

hbagdi commented on kong/schema_service_test.go:
```diff
@@ -0,0 +1,23 @@
+	assert.Nil(err)
+	assert.NotNil(client)
+
+	entities := []string{"services", "routes", "targets", "upstreams", "plugins"}
```
add all core entities, consult documentation or kong/db/schemas/entities dir in kong/kong repo

GGabriele replied:
Done, although I had to skip some entities, as `clustering_data_planes` and `parameters` are not present in < `2.3` and `workspaces` are only EE.

We should probably have a way to conditionally include or exclude test cases based on the running Kong version.

hbagdi replied:
> Done, although I had to skip some entities, as clustering_data_planes and parameters are not present in < 2.3 and workspaces are only EE.

No problem.

> We should probably have a way to conditionally include or exclude test cases based on the running Kong version.

We do. See runWhen* funcs.

hbagdi commented on kong/utils.go:
```diff
@@ -208,6 +214,167 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+}
+
+// FillUpstreamsDefaults ingests upstreams' defaults from their schema.
+func FillUpstreamsDefaults(upstream *Upstream, schema map[string]interface{}) error {
```
Since the code doesn't actually deal with the type of the first argument, can we have a generic function that takes in an "interface{}" and does the right thing based on type of the input.
You will need to use a "type switch" in Go.

GGabriele replied:
🧙‍♂️

hbagdi commented on kong/utils.go:
```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
 // FillPluginsDefaults ingests plugin's defaults from its schema.
 // Takes in a plugin struct and mutate it in place.
-func FillPluginsDefaults(plugin *Plugin, schema map[string]interface{}) error {
+func FillPluginsDefaults(plugin *Plugin, schema Schema) error {
```
We don't return Schema from plugin.GetFullSchema. Can we change that too?

GGabriele replied:
sure

hbagdi commented on kong/utils.go:
```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+	case *Upstream:
+		tmpEntity = &Upstream{}
+	}
+	defaults, err := getDefaultsObj(schema)
```
check for schema == nil before using

GGabriele replied:
Mmmh what do you think it's best here if `schema == nil`, returning `nil` or erroring out complaining about it being `nil`? Maybe the former?

hbagdi replied:
Return an error. When in doubt, return an error to the caller. That leads to better UX because the caller can then appropriate action. If the code returns without an error, the caller may assume that the function call worked only to be surprised later that it didn't.

GGabriele replied:
roger that

hbagdi commented on kong/utils.go:
```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+	}
+	defaults, err := getDefaultsObj(schema)
+	if err != nil {
+		return err
```
Annotate error,
```suggestion
		return fmt.Errorf("parse schema for defaults: %v", err)
```

hbagdi commented on kong/utils.go:
```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+		return err
+	}
+	if err = json.Unmarshal(defaults, &tmpEntity); err != nil {
+		return err
```
Annotate error.

hbagdi commented on kong/utils.go:
```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+	if err = json.Unmarshal(defaults, &tmpEntity); err != nil {
+		return err
+	}
+	return mergo.Merge(entity, tmpEntity)
```
Annotate error

hbagdi commented on kong/utils.go:
```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+	jsonSchemaWithDefaults, err := json.Marshal(&defaults)
+	if err != nil {
+		return nil, err
+	}
```
Code does a JSON marshal and unmarshal of defaults to get it into the typed structs. This can be fairly expensive, especially in hot code paths. The alternative is to use reflections.
No change is needed in the patch, only commenting to keep this in mind for the future.

hbagdi commented on kong/utils_test.go:
```diff
@@ -159,3 +161,356 @@ func Test_requestWithHeaders(t *testing.T) {
+		expected *Route
+	}{
+		{
+			name: "name paths",
```
```suggestion
			name: "fills defaults for all fields except paths, leaves name as is",
```
Descriptive tests, please.

hbagdi reviewed (commented):
Nice patch!

hbagdi commented on kong/utils_test.go:
```diff
@@ -159,3 +161,356 @@ func Test_requestWithHeaders(t *testing.T) {
+			},
+		},
+		// TODO: re-add once upstream patch has been merged.
+		// https://github.com/imdario/mergo/pull/203
```
It is likely that we will forget about this TODO. Can you remove the code and open a Github issue to track it?

GGabriele replied:
https://github.com/Kong/go-kong/issues/122

hbagdi commented:
@rainest Can I get a review from you on this one as well?

hbagdi reviewed (approved):
Let's move ahead on this one. If there are more issues, @GGabriele please take care of resolving with subsequent PRs.
